storagemarket/impl/clientutils/clientutils.go

```diff
@@ -5,37 +5,69 @@ import (
 	"context"
 
 	"github.com/ipfs/go-cid"
+	"github.com/ipld/go-car"
 	"github.com/multiformats/go-multibase"
 	"golang.org/x/xerrors"
 
 	"github.com/filecoin-project/go-address"
 	cborutil "github.com/filecoin-project/go-cbor-util"
-	"github.com/filecoin-project/go-commp-utils/pieceio"
-	"github.com/filecoin-project/go-multistore"
+	"github.com/filecoin-project/go-commp-utils/writer"
 	"github.com/filecoin-project/go-state-types/abi"
 	"github.com/filecoin-project/go-state-types/crypto"
 
 	"github.com/filecoin-project/go-fil-markets/shared"
 	"github.com/filecoin-project/go-fil-markets/storagemarket"
 	"github.com/filecoin-project/go-fil-markets/storagemarket/network"
+	"github.com/filecoin-project/go-fil-markets/stores"
 )
 
 // CommP calculates the commP for a given dataref
-func CommP(ctx context.Context, pieceIO pieceio.PieceIO, rt abi.RegisteredSealProof, data *storagemarket.DataRef, storeID *multistore.StoreID) (cid.Cid, abi.UnpaddedPieceSize, error) {
+// In Markets, CommP = PieceCid.
+// We can't rely on the CARv1 payload in the given CARv2 file being deterministic as the client could have
+// written a "non-deterministic/unordered" CARv2 file.
+// So, we need to do a CARv1 traversal here by giving the traverser a random access CARv2 blockstore that wraps the given CARv2 file.
+func CommP(ctx context.Context, carPath string, data *storagemarket.DataRef) (cid.Cid, abi.UnpaddedPieceSize, error) {
+	// if we already have the PieceCid, there's no need to do anything here.
 	if data.PieceCid != nil {
 		return *data.PieceCid, data.PieceSize, nil
 	}
 
+	// It's an error if we don't already have the PieceCid for an offline deal i.e. manual transfer.
 	if data.TransferType == storagemarket.TTManual {
 		return cid.Undef, 0, xerrors.New("Piece CID and size must be set for manual transfer")
 	}
 
-	commp, paddedSize, err := pieceIO.GeneratePieceCommitment(rt, data.Root, shared.AllSelector(), storeID)
+	if carPath == "" {
+		return cid.Undef, 0, xerrors.New("need Carv2 file path to get a read-only blockstore")
+	}
+
+	// Open a read-only blockstore off the CAR file, wrapped in a filestore so
+	// it can read file positional references.
+	fs, err := stores.ReadOnlyFilestore(carPath)
```

Comment:
I put a more in depth comment on the provider side, but I suspect there is an optimization here as well to avoid reading from a car then writing that car then reading to calculate commP. I think the optimization can be deferred but it is worth pursuing.

Reply:
@dirkmc follow-up here?

Reply:
@raulk please take a look 👁️👁️